File: packages/core-ts/src/signing.ts
import * as ed25519 from '@noble/ed25519';
import type { Keypair, AgentManifest, SignedManifest, SignedMessage } from './types.js';
import { canonicalizeManifest } from './manifest.js';
import { getAgentId } from './identity.js';
import { encodeBase64 } from './utils.js';

/** Sign a manifest, producing a signed envelope. */
export async function signManifest(
  manifest: AgentManifest,
  keypair: Keypair
): Promise<SignedManifest> {
  const payload = canonicalizeManifest(manifest);
  const payloadBytes = new TextEncoder().encode(payload);
  const signatureBytes = await ed25519.signAsync(payloadBytes, keypair.privateKey);

  return {
    payload,
    signature: encodeBase64(signatureBytes),
    key_id: getAgentId(keypair.publicKey),
    signed_at: new Date().toISOString(),
  };
}

/** Sign an arbitrary message (used for nonce-based handshakes). */
export async function signMessage(
  payload: string,
  nonce: string,
  keypair: Keypair,
  manifestRef: string
): Promise<SignedMessage> {
  const message = `${nonce}:${payload}:${manifestRef}`;
  const messageBytes = new TextEncoder().encode(message);
  const signatureBytes = await ed25519.signAsync(messageBytes, keypair.privateKey);

  return {
    payload: encodeBase64(new TextEncoder().encode(payload)),
    nonce,
    agent_id: getAgentId(keypair.publicKey),
    manifest_ref: manifestRef,
    signature: encodeBase64(signatureBytes),
    signed_at: new Date().toISOString(),
  };
}

File: packages/core-py/agentid/signing.py
from __future__ import annotations

import base64
from datetime import datetime, timezone

from nacl.signing import SigningKey

from agentid.identity import get_agent_id
from agentid.manifest import canonicalize_manifest
from agentid.types import AgentManifest, Keypair, SignedManifest, SignedMessage


def sign_manifest(manifest: AgentManifest, keypair: Keypair) -> SignedManifest:
    payload = canonicalize_manifest(manifest)
    signing_key = SigningKey(keypair.private_key)
    signed = signing_key.sign(payload.encode("utf-8"))
    signature = base64.b64encode(signed.signature).decode("ascii")

    return SignedManifest(
        payload=payload,
        signature=signature,
        key_id=get_agent_id(keypair.public_key),
        signed_at=datetime.now(timezone.utc).isoformat(),
    )


def sign_message(
    payload: str,
    nonce: str,
    keypair: Keypair,
    manifest_ref: str,
) -> SignedMessage:
    message = f"{nonce}:{payload}:{manifest_ref}"
    signing_key = SigningKey(keypair.private_key)
    signed = signing_key.sign(message.encode("utf-8"))
    signature = base64.b64encode(signed.signature).decode("ascii")

    return SignedMessage(
        payload=base64.b64encode(payload.encode("utf-8")).decode("ascii"),
        nonce=nonce,
        agent_id=get_agent_id(keypair.public_key),
        manifest_ref=manifest_ref,
        signature=signature,
        signed_at=datetime.now(timezone.utc).isoformat(),
    )
